### PR TITLE
hawkBit repository uses Optional on single entity find/get requests

### PR DESCRIPTION
**Pull request by kaizimmerm**
Another step of https://github.com/eclipse/hawkbit/issues/197

* Introduce Optional on find/get single results
* Adapt exception handling that EntityNotFound is not thrown on get/find requests form the repository but of necessary on caller side.
* While I was trying to figure out while our FEST assert where not supporting Optional I saw that we are are using a deprecated version (i.e. original FEST project that was stopped in 2013). Switched to the current version (assertj) that is in fact part of spring boot test.

### CODE REVIEW

File: hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetResource.java [hunk 1]
```diff
@@ -213,7 +214,9 @@
     public ResponseEntity<Void> cancelAction(@PathVariable("controllerId") final String controllerId,
             @PathVariable("actionId") final Long actionId,
             @RequestParam(value = "force", required = false, defaultValue = "false") final boolean force) {
-        final Action action = findActionWithExceptionIfNotFound(actionId);
+        final Action action = deploymentManagement.findAction(actionId)
+                .orElseThrow(() -> new EntityNotFoundException(Action.class, actionId));
+        ;
```
Comment (schabdo) [inline hawkbit-mgmt-resource/src/main/java/org/eclipse/hawkbit/mgmt/rest/resource/MgmtTargetResource.java]: remove `;`

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutGroupManagement.java [hunk 2]
```diff
@@ -131,14 +131,21 @@ public RolloutGroup findRolloutGroupById(final Long rolloutGroupId) {
...
+
+        final JpaRolloutGroup jpsRolloutGroup = (JpaRolloutGroup) rolloutGroup.get();
+
         final List<TotalTargetCountActionStatus> rolloutStatusCountItems = actionRepository
                 .getStatusCountByRolloutGroupId(rolloutGroupId);
 
         final TotalTargetCountStatus totalTargetCountStatus = new TotalTargetCountStatus(rolloutStatusCountItems,
-                Long.valueOf(rolloutGroup.getTotalTargets()));
-        rolloutGroup.setTotalTargetCountStatus(totalTargetCountStatus);
+                Long.valueOf(jpsRolloutGroup.getTotalTargets()));
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutGroupManagement.java]: typo of variable name `jpsRolloutGroup` => `jpaRolloutGroup` ?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaSoftwareManagement.java [hunk 3]
```diff
@@ -197,14 +192,18 @@ public SoftwareModule createSoftwareModule(final SoftwareModuleCreate c) {
...
+    public Optional<SoftwareModule> findSoftwareModuleById(final Long id) {
+        return Optional.ofNullable(softwareModuleRepository.findOne(id));
     }
 
     @Override
-    public SoftwareModule findSoftwareModuleByNameAndVersion(final String name, final String version,
+    public Optional<SoftwareModule> findSoftwareModuleByNameAndVersion(final String name, final String version,
             final Long typeId) {
 
+        if (!softwareModuleTypeRepository.exists(typeId)) {
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaSoftwareManagement.java]: This code-block is used 7 times in this class, maybe refactor this in a separate method.
```
if (!softwareModuleTypeRepository.exists(typeId)) {
   throw new EntityNotFoundException(SoftwareModuleType.class, typeId);
}
```

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 4]
```diff
@@ -313,6 +328,10 @@ public Long countTargetByFilters(final Collection<TargetUpdateStatus> status, fi
             specList.add(TargetSpecifications.isOverdue(TimestampCalculator.calculateOverdueTimestamp()));
         }
         if (filterParams.getFilterByDistributionId() != null) {
+            if (!distributionSetRepository.exists(filterParams.getFilterByDistributionId())) {
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: I think you can call the method `throwEntityNotFoundIfDsDoesNotExist` from line 258 here, is the same exists check, right?

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/RolloutManagement.java [hunk 5]
```diff
@@ -303,18 +302,19 @@ Rollout createRollout(@NotNull RolloutCreate rollout, @NotNull List<RolloutGroup
      *         does not exists
      */
     @PreAuthorize(SpringEvalExpressions.HAS_AUTH_ROLLOUT_MANAGEMENT_READ)
-    Rollout findRolloutByName(@NotNull String rolloutName);
+    Optional<Rollout> findRolloutByName(@NotNull String rolloutName);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/RolloutManagement.java]: `@NotNull` or `@NotEmpty`?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java [hunk 6]
```diff
@@ -922,41 +894,48 @@ public Rollout findRolloutByName(final String rolloutName) {
...
         update.getActionType().ifPresent(rollout::setActionType);
         update.getForcedTime().ifPresent(rollout::setForcedTime);
         update.getStartAt().ifPresent(rollout::setStartAt);
         update.getSet().ifPresent(setId -> {
-            final DistributionSet set = distributionSetManagement.findDistributionSetById(setId);
-            if (set == null) {
-                throw new EntityNotFoundException("Distribution set cannot be set as it does not exists" + setId);
-            }
+            final DistributionSet set = distributionSetManagement.findDistributionSetById(setId).orElseThrow(
+                    () -> new EntityNotFoundException("Distribution set cannot be set as it does not exists" + setId));
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java]: Use other EntityNotFoundException constructor with class

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java [hunk 7]
```diff
@@ -922,41 +894,48 @@ public Rollout findRolloutByName(final String rolloutName) {
...
+
             rollout.setDistributionSet(set);
         });
 
         return rolloutRepository.save(rollout);
     }
 
+    private JpaRollout getRolloutAndThrowExceptionIfNotFound(final Long rolloutId) {
+        return rolloutRepository.findById(rolloutId)
+                .orElseThrow(() -> new EntityNotFoundException("Rollout with id " + rolloutId + " not found."));
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java]: Use another EntityNotFoundException constructor with class

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java [hunk 8]
```diff
@@ -980,9 +958,8 @@ private void setRolloutStatusDetails(final Slice<JpaRollout> rollouts) {
 
     @Override
     public float getFinishedPercentForRunningGroup(final Long rolloutId, final Long rolloutGroupId) {
-        final RolloutGroup rolloutGroup = Optional.ofNullable(rolloutGroupRepository.findOne(rolloutGroupId))
-                .orElseThrow(() -> new EntityNotFoundException(
-                        "Rollout group with given ID " + rolloutGroupId + " not found."));
+        final RolloutGroup rolloutGroup = rolloutGroupRepository.findById(rolloutGroupId).orElseThrow(
+                () -> new EntityNotFoundException("Rollout group with given ID " + rolloutGroupId + " not found."));
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaRolloutManagement.java]: Use another EntityNotFoundException constructor with class

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 9]
```diff
@@ -110,23 +110,24 @@
...
-        final Target result = targetRepository.findByControllerId(controllerId);
+    public Optional<Target> findTargetByControllerIDWithDetails(final String controllerId) {
+        final Optional<Target> result = targetRepository.findByControllerId(controllerId);
         // load lazy relations
-        if (result != null) {
-            result.getTargetInfo().getControllerAttributes().size();
-            if (result.getTargetInfo() != null && result.getTargetInfo().getInstalledDistributionSet() != null) {
-                result.getTargetInfo().getInstalledDistributionSet().getName();
-                result.getTargetInfo().getInstalledDistributionSet().getModules().size();
+        if (result.isPresent()) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: Would prefer early return ;)

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java [hunk 10]
```diff
@@ -159,10 +160,8 @@ public Long countTargetsAll() {
 
     @Override
     public Slice<Target> findTargetsByTargetFilterQuery(final Long targetFilterQueryId, final Pageable pageable) {
-        final TargetFilterQuery targetFilterQuery = Optional
-                .ofNullable(targetFilterQueryRepository.findOne(targetFilterQueryId))
-                .orElseThrow(() -> new EntityNotFoundException(
-                        "TargetFilterQuery with given ID" + targetFilterQueryId + " not found"));
+        final TargetFilterQuery targetFilterQuery = targetFilterQueryRepository.findById(targetFilterQueryId)
+                .orElseThrow(EntityNotFoundException::new);
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java]: Use another EntityNotFoundException constructor with class

File: hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java [hunk 11]
```diff
@@ -327,12 +327,12 @@ private Action checkActionExist(final Message message, final ActionUpdateStatus
...
-            final Action findActionWithDetails = controllerManagement.findActionWithDetails(actionId);
-            if (findActionWithDetails == null) {
+            final Optional<Action> findActionWithDetails = controllerManagement.findActionWithDetails(actionId);
+            if (!findActionWithDetails.isPresent()) {
                 logAndThrowMessageError(message,
                         "Got intermediate notification about action " + actionId + " but action does not exist");
             }
-            return findActionWithDetails;
+            return findActionWithDetails.get();
         } catch (@SuppressWarnings("squid:S1166") final EntityNotFoundException e) {
```
Comment (SirWayne) [inline hawkbit-dmf-amqp/src/main/java/org/eclipse/hawkbit/amqp/AmqpMessageHandlerService.java]: The catch block can never reached, or?

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java [hunk 12]
```diff
@@ -61,4 +63,31 @@ public EntityNotFoundException(final String message, final Throwable cause) {
     public EntityNotFoundException(final String message) {
```
Comment (SirWayne) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/exception/EntityNotFoundException.java]: Maybe you can set this constructor to protected.

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java [hunk 13]
```diff
@@ -287,7 +287,7 @@ public String getStyle(final CellReference cellReference) {
         @Override
         public void click(final RendererClickEvent event) {
             rolloutUIState.setRolloutGroup(
-                    rolloutGroupManagement.findRolloutGroupWithDetailedStatus((Long) event.getItemId()));
+                    rolloutGroupManagement.findRolloutGroupWithDetailedStatus((Long) event.getItemId()).get());
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/rollout/rolloutgroup/RolloutGroupListGrid.java]: What happen if the optional is not present?

**Comment by kaizimmerm** [comment]
@SirWayne I think I found a nice solution for getting rid `EntityNotFoundException(final String message)`

**Pull request merged**